src/java.base/unix/native/libnet/net_util_md.c

```diff
@@ -171,21 +171,35 @@ jint reuseport_supported(int ipv6_available)
 
 void NET_ThrowUnknownHostExceptionWithGaiError(JNIEnv *env,
                                                const char* hostname,
-                                               int gai_error)
+                                               int gai_error,
+                                               int sys_errno)
 {
     int size;
     char *buf;
-    const char *format = "%s: %s";
+    const char *format;
+    const char *sys_errno_string;
     const char *error_string = gai_strerror(gai_error);
-    if (error_string == NULL)
+    if (error_string == NULL) {
         error_string = "unknown error";
-
-    size = strlen(format) + strlen(hostname) + strlen(error_string) + 2;
+    }
+    if (gai_error == EAI_SYSTEM) {
+        // EAI_SYSTEM implies that the actual error is stored in the system errno.
+        // we thus additionally get the string representation for that system errno.
+        sys_errno_string = strerror(sys_errno);
+        format = "%s: %s: %s";
+        size = strlen(format) + strlen(hostname) + strlen(error_string) + strlen(sys_errno_string) + 2;
```

Comment:
You are right, although I'm not sure why there's a `strlen(format)` in that size computation. The original change was introduced in https://hg.openjdk.org/jdk7/jdk7/jdk/rev/b5d37597c815 and at that time it was using `sprintf` (and not `snprintf`) and the `sprintf` as per its documentation, considers size to be `INT_MAX + 1`, so `size` wasn't being passed to it and only used for `malloc()`. Even then, I don't understand why `strlen(format)` is considered for the size.

I have updated this part of the PR, but I'm going to look at this size computation bit more tomorrow with a fresh mind.

Reply:
oh - yes - the `strlen(format)` is strange - maybe it allowed to accout for the various ": ", and the rest would be > 1 which would allow for the null terminator? But then + 2 wouldn't have been needed.